### ISSUE TITLE
Lynx BMS: add /Alarms/LowCellVoltage

### ISSUE THREAD

**Issue by mpvader** [victronenergy/venus#357]
- [x] Add it do the vecan-dbus  driver
- [x] Add it to the gui (warning & alarm)
- [x] Add it to ModbusTCP
- [x] Add it to Vrmlogger

### Specification

2019-1-11: Updated this to reflect outcome of below discussions.

/Alarms/LowCellVoltage, which can have the following values:

0 - No alarm
1 - Warning: almost discharged, based on bit 3 of VE_REG_BMS_FLAGS = Almost discharged
2 - Alarm: fully discharged, based on bit 2 of VE_REG_BMS_FLAGS = VE_INTERFACE_BMS_FLAGS_DISCHARGED.

The alarm is raised at the same time that the BMS has disabled the 'allow-to-discharge' signal; which shuts down any connected inverters and other loads.

### Original text

do we raise any alarms on that already? Is it on dbus? Looks like its not:

```
root@ccgx:~# dbus -y com.victronenergy.battery.socketcan_can0_di1_uc589827 / GetText
value = {
 'Balancing': '0',
 'Capacity': '100Ah',
 'Connected': '1',
 'ConsumedAmphours': '0.0Ah',
 'Dc/0/Current': '28.0A',
 'Dc/0/Power': '777W',
 'Dc/0/Temperature': '30 degrees',
 'Dc/0/Voltage': '27.77V',
 'DeviceInstance': '1',
 'Diagnostics/LastErrors/1/Error': '13',
 'Diagnostics/LastErrors/1/Time': '1537430766',
 'Diagnostics/LastErrors/2/Error': '0',
 'Diagnostics/LastErrors/2/Time': '0',
 'Diagnostics/LastErrors/3/Error': '0',
 'Diagnostics/LastErrors/3/Time': '0',
 'Diagnostics/LastErrors/4/Error': '0',
 'Diagnostics/LastErrors/4/Time': '0',
 'Diagnostics/ShutDownsDueError': '',
 'ErrorCode': '0',
 'FirmwareVersion': 'v1.13',
 'HardwareVersion': '',
 'History/AutomaticSyncs': '18',
 'History/AverageDischarge': '',
 'History/CanBeCleared': '1',
 'History/ChargeCycles': '',
 'History/ChargedEnergy': '72.57kWh',
 'History/Clear': '',
 'History/DeepestDischarge': '376.8Ah',
 'History/DischargedEnergy': '60.67kWh',
 'History/FullDischarges': '',
 'History/LastDischarge': '',
 'History/MaximumCellVoltage': '4.15V',
 'History/MaximumTemperature': '36 degrees',
 'History/MaximumVoltage': '29.17V',
 'History/MinimumCellVoltage': '2.91V',
 'History/MinimumTemperature': '19 degrees',
 'History/MinimumVoltage': '21.40V',
 'History/TimeSinceLastFullCharge': '',
 'History/TotalAhDrawn': '2376.1Ah',
 'Info/BatteryLowVoltage': '21.00V',
 'Info/MaxChargeCurrent': '30.0A',
 'Info/MaxChargeVoltage': '28.56V',
 'Info/MaxDischargeCurrent': '100.0A',
 'Io/AllowToCharge': '1',
 'Io/AllowToDischarge': '1',
 'Io/ExternalRelay': '',
 'Io/ProgrammableContact': '0',
 'Mgmt/Connection': 'VE.Can',
 'Mgmt/ProcessName': 'vecan-dbus',
 'Mgmt/ProcessVersion': '2.28',
 'N2kUniqueNumber': '589827',
 'ProductId': '0xA392',
 'ProductName': 'Lynx Ion BMS 400A',
 'Serial': '',
 'Settings/HasSettings': '0',
 'Settings/HasTemperature': '1',
 'Soc': '100.0%',
 'State': '9',
 'System/BatteriesParallel': '1',
 'System/BatteriesSeries': '1',
 'System/MaxCellTemperature': '30 degrees',
 'System/MaxCellVoltage': '3.96V',
 'System/MinCellTemperature': '30 degrees',
 'System/MinCellVoltage': '3.95V',
 'System/NrOfBatteries': '1',
 'System/NrOfCellsPerBattery': '0',
 'SystemSwitch': '1',
 'TimeToGo': '176040s'}
```

**Comment by mpvader** [comment]
Its a flag in the same VREG that has the AllowToCharge & Discharge flag, says Wilco

**Comment by martinbosma** [comment]
Right now only `/ErrorCode` triggers an alarm for all Lynx BMSes. I can add an alarm on "almost discharged" quite easily. This requires modification of `gui` and `debus-vecan`.

I propose to add `/Alarms/AlmostDischarged` which reflects bit 3 of `VE_REG_BMS_FLAGS`,

@mpvader What is the priority of this?

**Comment by mpvader** [comment]
its not urgent, but also just a small task. So pls do in the coming few weeks when you want a break of a larger project.

**Comment by martinbosma** [comment]
@mpvader Should the `Alarms/AlmostDischarged` have the value `1` (warning) or `2` (alarm) on the dbus? Or is vrm-logger unable to cope with warnings and does it always expect the value `2`? Either way is fine for me.

**Comment by martinbosma** [comment]
As discussed with @mpvader:
Instead of `Alarms/AlmostDischarged` we will make `Alarms/LowCellVoltage`, which can have the following values:
- `0` - No alarm
- `1` - Warning: almost discharged, based on bit 3 of `VE_REG_BMS_FLAGS`
- `2` - Alarm: fully discharged, based on bit 2 of `VE_REG_BMS_FLAGS`

This will work in case of the (new) Lynx Ion BMS, because bit 2 & 3 are determined on cell voltage.

For the old BMSes, we have a different VREG (`VE_REG_LYNX_ION_FLAGS`), which is almost the same as `VE_REG_BMS_FLAGS`. However, the definition of bit 2 (fully discharged), actually reflect the inverted "allow to discharge" signal, which is not only based on cell voltage, **but also on temperature.**

We can do 2 things:
1. Only support the warning (almost discharged) for these older BMSes. The new BMS will have both warning and alarm.
2. Only support the warning (almost discharged) for all BMSes. Perhaps even go back to the `Alarms/AlmostDischarged` proposal.

**Comment by mpvader** [comment]
Only support the path in new BMS-es. We don’t sell the old one anymore.

**Comment by martinbosma** [comment]
Does the path `Alarms/LowCellVoltage` also need to be added to others services? E.g. vrm-logger or modbus-tcp?

**Comment by izak** [comment]
Included in Venus v2.30~28.
